### PR TITLE
Pulumi storage fixes, including adding encryption key

### PR DESCRIPTION
**Pull request by craddm**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).
- [ ] You have marked this pull request as a **draft** and added `'[WIP]'` to the title if needed (if you're not yet ready to merge).

### :arrow_heading_up: Summary

Adds additional fixes from co-working session, and preliminary attempts to include encryption key in `pulumi.yaml`

### CODE REVIEW

File: data_safe_haven/commands/context.py
Comment (JimMadge) [inline data_safe_haven/commands/context.py]: We should add a test for the case where there is no context.

File: data_safe_haven/commands/deploy.py
Comment (JimMadge) [inline data_safe_haven/commands/deploy.py:80]: I can't remember if there are tests for the deploy command but we should add one to check that upload is always called whether or not an exception is raised.

File: data_safe_haven/config/pulumi.py
Comment (JimMadge) [inline data_safe_haven/config/pulumi.py:13]: Again, need to update tests for this change.

**Review by JimMadge — COMMENTED** [review]
All good changes, do you want to get these merged into `pulumi_storage` or shall we work on this branch until we have fixed all issues? @craddm @jemrobinson

**Comment by jemrobinson** [comment]
@JimMadge let's merge this in (once your comments are addressed). If we need to make more changes we can do that on a new branch.

File: data_safe_haven/commands/context.py
Comment (jemrobinson) [reply data_safe_haven/commands/context.py]: Can you add this or open an issue? I think you're most familiar with this set of changes and the test suite.

File: data_safe_haven/commands/deploy.py
Comment (jemrobinson) [reply data_safe_haven/commands/deploy.py:80]: Can you add this or open an issue?

**Comment by craddm** [comment]
Are we just merging this in before getting the encrypted key thing working, and then getting it working on the main `pulumi_storage` branch?

**Comment by JimMadge** [comment]
With regards to opening issues, I'm not really comfortable deferring tests or merging changes which break PR. I don't think there is any great rush here so we should do that before merging.

**Comment by jemrobinson** [comment]
Also agree with Matt that we should get the encrypted key piece working. Are you able to test this @JimMadge ?

**Comment by JimMadge** [comment]
Good point, we should check that change is doing what we hope :+1:. If not we can cherry-pick or merge the other changes anyway.

I'll try to find time to test this. This set of PRs are my only priority when it comes to development.

File: data_safe_haven/commands/context.py
Comment (JimMadge) [reply data_safe_haven/commands/context.py]: fixed in a3d9eba2686240f73964dbafa13a14ddbaaa24f0 and fb280b9536e12fcd82bf9c3624f8f97f9ed619b4

File: data_safe_haven/config/pulumi.py
Comment (JimMadge) [reply data_safe_haven/config/pulumi.py:13]: done in 1f0b05a840abac019e46bf44d8d404981f8c7a2d and 42a11c0164f66e40d89ad43de464ed4975a15d76

File: data_safe_haven/commands/deploy.py
Comment (JimMadge) [reply data_safe_haven/commands/deploy.py:80]: Done in [db1fc3d](https://github.com/alan-turing-institute/data-safe-haven/pull/1839/commits/db1fc3d40bea75dfd850c4e7dd0cf1ed4a4277ae)

**Review by JimMadge — APPROVED** [review]

**Pull request merged**
